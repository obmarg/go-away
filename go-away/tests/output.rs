#![allow(dead_code)]

use insta::assert_snapshot;

use go_away::{TypeMetadata, TypeRegistry};

#[derive(TypeMetadata)]
struct MyData {
    field_one: String,
    nested: Nested,
}

#[test]
fn test_struct_output() {
    let mut registry = TypeRegistry::new();
    MyData::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum NewTypeEnum {
    OptionOne(One),
    OptionTwo(Two),
}

#[derive(TypeMetadata)]
struct Nested {
    #[serde(rename = "some_other_name")]
    a_string: String,
    an_int: i64,
    fulfilment_type: FulfilmentType,
}

#[derive(TypeMetadata)]
enum FulfilmentType {
    Delivery,
    Collection,
}

#[derive(TypeMetadata)]
struct One {
    x: f32,
}

#[derive(TypeMetadata)]
struct Two {
    y: bool,
}

#[test]
fn test_newtype_enum() {
    let mut registry = TypeRegistry::new();
    NewTypeEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type", content = "data")]
enum StructEnum {
    OptionOne { x: String, y: i32 },
    OptionTwo { foo: String, bar: Nested },
}

#[test]
fn test_struct_enum() {
    let mut registry = TypeRegistry::new();
    StructEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}

#[derive(TypeMetadata)]
#[serde(tag = "type")]
enum InternallyTaggedTupleEnum {
    One(One),
    Two(Two),
}

#[test]
fn test_internally_tagged_tuple_enum() {
    let mut registry = TypeRegistry::new();
    InternallyTaggedTupleEnum::metadata(&mut registry);

    assert_snapshot!(go_away::registry_to_output(registry));
}
