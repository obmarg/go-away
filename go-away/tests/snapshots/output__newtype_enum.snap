---
source: go-away/tests/output.rs
expression: "go_away::registry_to_output(registry)"

---
type Two struct {
	Y bool `json:"y"`
}
type One struct {
	X float64 `json:"x"`
}
type NewTypeEnum struct {
	OptionOne *One
	OptionTwo *Two
}

func (self NewTypeEnum) MarshalJSON() ([]byte, error) {
	if err := self.Validate(); err != nil {
		return nil, fmt.Errorf("Validate Failed: %w", err)
	}
	if self.OptionOne != nil {
		output := make(map[string]interface{})
		output["type"] = "OptionOne"
		output["data"] = self.OptionOne
		return json.Marshal(output)
	} else 	if self.OptionTwo != nil {
		output := make(map[string]interface{})
		output["type"] = "OptionTwo"
		output["data"] = self.OptionTwo
		return json.Marshal(output)
	} else 	{
		return nil, fmt.Errorf("No variant was present")
	}
}
func (self *NewTypeEnum) UnmarshalJSON(data []byte) error {
	temp := struct {
		Tag string `json:"type"`
	}{}
	if err := json.Unmarshal(data, &temp); err != nil {
		return err
	}
	if temp.Tag == "OptionOne" {
		rv := struct {
			Data One `json:"data"`
		}{}
		if err := json.Unmarshal(data, &rv); err != nil {
			return err
		}
		self.OptionOne = &rv.Data
		self.OptionTwo = nil
	} else if temp.Tag == "OptionTwo" {
		rv := struct {
			Data Two `json:"data"`
		}{}
		if err := json.Unmarshal(data, &rv); err != nil {
			return err
		}
		self.OptionTwo = &rv.Data
		self.OptionOne = nil
	} else {
		return errors.New("Unknown type tag")
	}
	return nil
}
func (u NewTypeEnum) Validate() error {
	var count int

	if u.OptionOne != nil {
		count++
	}

	if u.OptionTwo != nil {
		count++
	}

	if count != 1 {
		return fmt.Errorf("one variant must be populated, found %d", count)
	}

	return nil
}

